cns/restserver/internalapi.go

```diff
@@ -378,9 +380,11 @@ func (service *HTTPRestService) ReconcileIPAMState(ncReqs []*cns.CreateNetworkCo
 		}
 	}
 
-	if err := service.MarkExistingIPsAsPendingRelease(nnc.Spec.IPsNotInUse); err != nil {
-		logger.Errorf("[Azure CNS] Error. Failed to mark IPs as pending %v", nnc.Spec.IPsNotInUse)
-		return types.UnexpectedError
+	if nnc != nil {
```

Comment:
Curious, how did this work before? If nnc == nil currently, L384 would have panicked.

Reply:
It is currently never nil. But it will be in the future, in another PR that I plan to open after this one. In nodesubnet, we don't use NNCs.

Reply:
there is only one caller which doesn't allow a nil NNC to proceed
https://github.com/Azure/azure-container-networking/blob/79c2af56a536fff2205a1c90521f20616494708e/cns/service/main.go#L1134-L1143
I'm not opposed to re-checking in this func since we're passing a pointer around, but a nil NNC should be a fatal error and we should _immediately_ validate the inputs instead of getting mostly done with the work before we check it. 
Passing a nil ref is not a sustainable strategy for configuring behavior. It's hard to discover, except accidentally when we expect one behavior and are getting another because we _accidentally_ have a nil and should have errored earlier instead.

Reply:
@rbtr , As you know, in Nodesubnet, we won't have any NNCs created, so this change is really to facilitate the use of this method to reconcile IPAM state for Nodesubnet, when no NNCs exist. Now, given we don't want nil as a parameter, I could pass an "empty" NNC, and then replace this check with a check on the `ncReq`, so that we don't end up trying to do NNC-specific  things in nodesubnet. Or I could just create a `ReconcileIPAMStateForNodesubnet` method that'd be substantially similar to this function, but without the NNC parameter. Shall I go with the empty NNC approach?

Reply:
@rbtr, thoughts?

Reply:
The function is now broken into smaller functions, and we don't use the nil parameter anymore. Resolving this thread to allow merge to proceed. We have a todo in the code to address any further feedback.